File: app/src/main/java/com/example/dreamloaf/viewmodels/ProductionViewModel.java
package com.example.dreamloaf.viewmodels;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.example.dreamloaf.data.Production;
import com.example.dreamloaf.repository.ProductionRepository;

import java.util.List;
import java.util.concurrent.Executors;

public class ProductionViewModel extends AndroidViewModel {
    private final ProductionRepository repository;
    private final LiveData<List<Production>> allProduction;

    public ProductionViewModel(@NonNull Application application) {
        super(application);
        repository = new ProductionRepository(application);
        allProduction = repository.getAllProduction();
    }

    public LiveData<List<Production>> getProductionByDate(String date) {
        return repository.getProductionByDate(date);
    }

    public void addProduction(Production production) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Production existing = repository.getProductionForProductAndDate(
                    production.getProductId(),
                    production.getDate()
            );

            if (existing != null) {
                existing.setQuantity(production.getQuantity());
                repository.update(existing);
            } else {
                repository.insert(production);
            }
        });
    }

    public void saveProductionData(int productId, int quantity, String date) {
        repository.saveProductionData(productId, quantity, date);
    }

    public void updateProduction(Production production) {
        repository.update(production);
    }

    public void deleteProduction(Production production) {
        repository.delete(production.getId());
    }

    public LiveData<List<Production>> getAllProduction() {
        return allProduction;
    }
}

File: app/src/main/java/com/example/dreamloaf/viewmodels/ProductionViewModel.kt
package com.example.dreamloaf.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import com.example.dreamloaf.data.Production
import com.example.dreamloaf.repository.ProductionRepository
import java.util.concurrent.Executors

class ProductionViewModel(application: Application) : AndroidViewModel(application) {
    private val repository: ProductionRepository
    val allProduction: LiveData<MutableList<Production?>?>?

    init {
        repository = ProductionRepository(application)
        allProduction = repository.allProduction
    }

    fun getProductionByDate(date: String?): LiveData<MutableList<Production?>?>? {
        return repository.getProductionByDate(date)
    }

    fun addProduction(production: Production) {
        Executors.newSingleThreadExecutor().execute {
            val existing = repository.getProductionForProductAndDate(
                production.productId,
                production.date
            )
            if (existing != null) {
                existing.quantity = production.quantity
                repository.update(existing)
            } else {
                repository.insert(production)
            }
        }
    }

    fun saveProductionData(productId: Int, quantity: Int, date: String?) {
        repository.saveProductionData(productId, quantity, date)
    }

    fun updateProduction(production: Production) {
        repository.update(production)
    }

    fun deleteProduction(production: Production) {
        repository.delete(production.id)
    }
}